Feature/php rt

beforeisland opened the pull request:
PHPテキストの掲示板アプリへのいいね機能及びリツイート機能の追加。

DBのテーブル構造は以下の通り。

1.membersテーブル(テキストから変更なし)

2.postsテーブル(リツイート機能用に以下のカラムを追加)
rt_post_id INT,
rt_member_id INT

3goodテーブル(いいね機能用に新規追加)
good id  INT AUTO_INCREMENT PRIMARY KEY,
member_id INT,
post_id INT,
created DATETIME,
modified TIMESTAMP

beforeisland committed:
add: いいね数の表示機能の実装

投稿ごとのいいね数を取得するSQLを発行し、post_idで呼び出す

beforeisland committed:
add: いいねボタンのON-OFF機能の実装

条件分岐によるいいね機能のON-OFFのリンクを実装。ログイン者か否か、その投稿にいいねが存在しているかをスイッチとした、

beforeisland committed:
fix: いいね取得のSQL修正

いいね取得方法を、投稿idに加えて第二引数にログイン者のidを指定することで他者のいいねを取り消せてしまう不具合を修正した

beforeisland committed:
add: いいねの削除機能を実装

goodテーブルのデータを削除することによって画面に表示されるいいねを削除する機能を実装

beforeisland committed:
Revert "add: いいねの削除機能を実装"

This reverts commit e6c4a994899dcc71db83e6da80b6308dbeb86382.

beforeisland committed:
fix: いいね削除機能の実装修正

コミットミスの修正

beforeisland committed:
add: いいね追加機能の実装

goodテーブルにINSERTで追加する形式によりいいね追加機能を実装しました

beforeisland committed:
Revert "add: いいね追加機能の実装"

This reverts commit a2c3f4072be7f8ddfdc349e8c92efcc339ccf6c6.

beforeisland committed:
fix: コミットミス修正

beforeisland committed:
fix: コミットミス修正

beforeisland committed:
add: リツイート投稿者表示機能実装

postsテーブルのrt_post_idが入っている投稿に対してそのリツイート者を画面表示する

beforeisland committed:
add: リツイート数の画面表示実装

リツイート投稿か元投稿か判断しつつリツイート数を表示する機能を実装

beforeisland committed:
add: リツイート投稿に対するいいね数カウント及びリツイート元投稿同期いいねON-OFFを実装

条件分岐によりリツイート投稿か否かを判断し実装

beforeisland committed:
add: RT機能のON-OFF実装

条件分岐によるリツイート追加：削除判断機能、リツイート投稿とリツイート元投稿のリツイートボタン連動を実装

beforeisland committed:
add: リツイート追加削除機能追加

urlパラメータを使ったリツイート追加削除機能を実装

beforeisland committed:
fix: リツイート機能の不具合修正

自分以外のリツイート投稿かつ未リツイートの投稿に対するリツイート機能が反応しなかった。原因はurlパラメータの指定変数が間違っていたことによるため修正。

beforeisland committed:
update: いいね-リツイートスイッチをアイコンに変更

いいねとリツイート機能を、文字列によるリンクをクリックすることによりON-OFFしていたものを、アイコンによるスイッチに変更。

beforeisland committed:
fix: リツイート投稿に対する削除リンク非表示に変更

リツイート投稿はリツイートアイコンで削除すればいいうえに、他人のリツイートを投稿元ツイートの投稿者が削除できてしまう挙動だったため、リツイート投稿については削除機能を非表示とした。

beforeisland committed:
fix: コメント修正

コードのコメントを修正

beforeisland committed:
add バックアップファイルを保存(SQL)

SQLのバックアップファイルをbackupfileフォルダに格納。
*テキストからのDB改造点
1.postsテーブルへのカラム追加)(リツイート機能用)
 rt_post_id, rt_member_idカラム追加
2.goodテーブルの追加(いいね機能用)

choimake commented on good_delete.php:
```diff
@@ -0,0 +1,22 @@
+require('dbconnect.php');
+
+if (isset($_SESSION['id'])) {
+	$good_id = $_REQUEST['good_id'];
+
+	//いいね投稿を検査する
+	$goods = $db->prepare('SELECT * FROM good WHERE good_id=?');
+	$goods->execute(array($good_id));
+	$good= $goods->fetch();
```
リクエストで渡された値をそのままSQLに入力していますが、セキュリティ的な観点で良くありません。
SQLに想定していない値（この場合でいえば「a」みないな文字列など）が入ることで想定外のバグを発生する可能性があるからです。

想定していない値がリクエストされた時（ここの場合は数字以外）は、SQLに入れる前にエラーになるような処理を追加してください。
（これをバリデーションといいます）

beforeisland replied:
修正完了。条件分岐にて不正な値(数字以外)が入った場合にエラーメッセージを表示する処理を追加しました。バリデーションを意識したコーディングをしていきます。

update: セッション・リクエストで受け取る数値のチェック機能を追加	78f129a	beforeisland <beforeisland@gmail.com>	2020/10/12 22:13

choimake commented on good_delete.php:
```diff
@@ -0,0 +1,22 @@
+
+header('Location: index.php');
+exit();
+?>
```
ファイルの末尾には改行を入れましょう。

Githubでファイル末尾に改行がないと、赤いマークが出ます。

何故かと言いますと、
「テキストファイルは本来（POSIX的には）必ず最後は改行コードで終わっていなければならない」というルールがあるからです。

それが守られていないときにGithubでは赤いマークという形で警告がでます。

以下、参考URLです。
https://www.cqpub.co.jp/interface/sample/200902/if02_038.pdf

標準規格やIEEE、POSIXなどについて一度確認してみてください。

beforeisland replied:
修正完了。html課題の際にも指摘があった項目であり、内容は認識しているものの意識が足りていなかった内容でした。今後気をつけて行きます。

fix: ファイル最終行へ改行挿入	7f327e4	beforeisland <beforeisland@gmail.com>	2020/10/11 20:18

choimake commented on good_delete.php:
```diff
@@ -0,0 +1,22 @@
+	$goods->execute(array($good_id));
+	$good= $goods->fetch();
+
+	if($good['member_id'] == $_SESSION['id']) {
+		//いいねを削除する
+		$del = $db->prepare('DELETE FROM good WHERE good_id=?');
+		$del->execute(array($good_id));
+	}
```
値の比較をするときは厳密な比較「===」「!==」などを利用するようにしてください。
型を意識したプログラミングができるようにしましょう。

beforeisland replied:
修正完了。PHPでは型宣言の必要がないがゆえに想定外の処理が起こりうることを把握しました。意識して厳密な比較ができる演算子を利用します。

fix: 値の比較演算子を厳密な比較ができるものに変更	fdbf4be	beforeisland <beforeisland@gmail.com>	2020/10/12 22:27

choimake commented on backupfile/mini_bbs.sql:
```diff
@@ -0,0 +1,164 @@
+-- テーブルの構造 `good`
+--
+
+CREATE TABLE `good` (
```
テーブル名が複数形と単数形が混ざっているので、どちらかに統一してください。
（この場合は複数形のテーブル数が多いので、複数形で良いかなと思います）

beforeisland replied:
修正完了。今後はデータベース設計時に意識していきます。

fix: テーブル名を複数形に統一	3a830f7	beforeisland <beforeisland@gmail.com>	2020/10/11 20:13

gacha1987 replied:
データベース更新したら、backupfile配下も更新してください。

beforeisland replied:
更新完了。

update: バックアップしたデータベースのファイルを更新	0edbfb0	beforeisland <beforeisland@gmail.com>	2020/10/15 0:57

choimake reviewed (commented):
コメントしました。
挙動としては全く問題ないです。全て動作要件を満たしています。

ただいくつかセキュリティ面、コーディング面、その他の点に関していくつか気になる点があるので、確認の上修正してください。

beforeisland committed:
fix: テーブル名を複数形に統一

goodテーブルのみ名称が単数形だったため、goodsに名称変更しました。併せてプログラム内のSQLを全て修正。

beforeisland committed:
fix: ファイル最終行へ改行挿入

テキストファイルは改行で終わる必要があるため修正

beforeisland committed:
update: セッション・リクエストで受け取る数値のチェック機能を追加

セキュリティ対策として、リツイート及びいいね機能におけるセッション・リクエストで受け取る値が数値かどうかをチェックする機能を追加。また、不正な値が入力された際にエラーメッセージが表示されるよう処理を追加。

beforeisland committed:
fix: 値の比較演算子を厳密な比較ができるものに変更

値の比較演算子を型から比較できるよう変更。「==」=>「===」

choimake commented on good_insert.php:
```diff
@@ -0,0 +1,17 @@
+    $insert = $db->prepare('INSERT INTO goods SET member_id=?, post_id=?, created=NOW()');
+    $insert->execute(array(
+        $_SESSION['id'],
+        $_REQUEST['post_id']
```
リクエストの値に関してバリデーションを実装してください。

beforeisland replied:
修正完了。前回依頼時に中途半端な状態でPR依頼をしてしまいました。失礼いたしました。

update: いいね・リツイートの追加機能用に渡すリクエストの検査機能を追加	d7f4139	beforeisland <beforeisland@gmail.com>	2020/10/13 22:56

choimake commented on good_insert.php:
```diff
@@ -0,0 +1,17 @@
+
+header('Location: index.php');
+exit();
+?>
```
前回の指摘漏れで申し訳ありませんが、phpのみのファイルに関しては、末尾に`?>`を付けない実装方法が推奨されています。
こちら削除をお願いします。

以下参考
https://thesaibase.com/php/end-tag

beforeisland replied:
修正完了。閉じタグは省略しなければならない件、把握しました。

fix: phpファイル末尾の閉じタグ削除	d80b448	beforeisland <beforeisland@gmail.com>	2020/10/13 23:02

choimake commented on rt_insert.php:
```diff
@@ -0,0 +1,23 @@
+if (is_numeric($_SESSION['id'])) {
+    //リツイートを追加する
+    $rtCopys = $db->prepare('SELECT * FROM posts WHERE id=?');
+    $rtCopys->execute(array($_REQUEST['post_id']));
```
ここのリクエストもバリデーションが必要です。

beforeisland replied:
修正完了。前回依頼時に中途半端な状態でPR依頼をしてしまいました。失礼いたしました。

update: いいね・リツイートの追加機能用に渡すリクエストの検査機能を追加	d7f4139	beforeisland <beforeisland@gmail.com>	2020/10/13 22:56

choimake reviewed (commented):
コメントしました。
前回の指摘漏れを多分に含んでおり、大変申し訳ありませんが、そちらも合わせて対応お願いします。

- 現在、images以下の画像ファイルがバージョン管理されている状態になっていますが、これは共有の必要がないはずです。こちらバージョン管理をしない設定に変更してください。

beforeisland committed:
fix: gitgnoreリストにimagesフォルダを追加

imagesフォルダ以下の画像ファイルはバージョン管理の必要性がないため、管理対象から外した

beforeisland committed:
fix: gitignoreリスト末尾へ改行追加

GitHub対応修正

beforeisland commented:
> コメントしました。
> 前回の指摘漏れを多分に含んでおり、大変申し訳ありませんが、そちらも合わせて対応お願いします。
> 
> * 現在、images以下の画像ファイルがバージョン管理されている状態になっていますが、これは共有の必要がないはずです。こちらバージョン管理をしない設定に変更してください。

バージョン管理をしない設定に変更しました。バージョン管理の効率化や容量削減に繋がるようですね。

fix: gitgnoreリストにimagesフォルダを追加	4e62d7e	beforeisland <beforeisland@gmail.com>	2020/10/13 22:42

beforeisland committed:
update: いいね・リツイートの追加機能用に渡すリクエストの検査機能を追加

セキュリティ対策にて、URLパラメータにてリクエストで渡す値が数値かどうかを判定するための処理及びエラーメッセージを追加

beforeisland committed:
fix: phpファイル末尾の閉じタグ削除

純粋なphpファイル末尾のphp閉じタグは省略しなければならないため削除

gacha1987 commented on good_delete.php:
```diff
@@ -0,0 +1,36 @@
+
+	} else {
+		echo 'error:URLパラメータが不正です';
+        exit();
```
exit()だけスペースで他はタブみたいなのでインデントを統一しましょう。

beforeisland replied:
修正完了。インデントについては、フォーマッターのインストール及びVScodeの設定変更により、スペースによる字下げへ統一をはかりました。

fix: インデント設定変更 a3b2a46 beforeisland beforeisland@gmail.com 2020/10/14 23:59

gacha1987 commented on good_insert.php:
```diff
@@ -0,0 +1,27 @@
+if(is_numeric($_SESSION['id'])) {
+
+    //if:リクエストにて受け取ったURLパラメータが数値形式で存在しているか調査
+    if(is_numeric($_REQUEST['post_id'])) {
```
バリデーションの観点として `いいねをする投稿がそもそも存在するか` というチェックを行うと、より品質の高いコードになります！
例えばpost_idが改竄され、存在しないpost_id(99とか)を指定された場合、数値なので現在のバリデーションは通過し、goodsテーブルに存在しない投稿に対してのいいねが登録され、エラーを引き起こす可能性が生まれるので、そういった箇所は潰していきましょう。

beforeisland replied:
機能追加。
ご教示ありがとうございます！URLパラメータで値を渡しす方式でプログラムを構築しましたが、簡単に弄れてしまうことを認識しました。

add: いいね追加機能にいいね対象投稿の存在有無の検知機能を追加	b668a23	beforeisland <beforeisland@gmail.com>	2020/10/15 0:37

beforeisland replied:
リツイート機能にも同様の機能を追加しました。

add: リツイート追加機能にリツイート対象投稿の存在有無の検知機能を追加	a25a0f3	beforeisland <beforeisland@gmail.com>	2020/10/15 0:53

gacha1987 commented on good_delete.php:
```diff
@@ -0,0 +1,36 @@
+			$del = $db->prepare('DELETE FROM goods WHERE good_id=?');
+			$del->execute(array($_REQUEST['good_id']));
+		} else {
+			echo 'error:ログイン者といいね投稿者が一致しません';
```
exit()がありません.

beforeisland replied:
修正完了。本件、処理の追記忘れです。

fix: いいね・リツイート消去プログラムへエラー処理追加	d000130	beforeisland <beforeisland@gmail.com>	2020/10/15 0:04

gacha1987 commented on index.php:
```diff
@@ -111,42 +138,162 @@ function makeLink($value) {
+                    if($_SESSION['id'] === $post['member_id']): ?>
+                    [<a href="delete.php?id=<?php echo h($post['id']); ?>" style="color:#F33;">削除</a>]
+                    <?php endif; 
+					}
```
タブとスペースが混在しているので統一してください。

beforeisland replied:
修正完了。インデントについては、フォーマッターのインストール及びVScodeの設定変更により、スペースによる字下げへ統一をはかりました。

fix: インデント設定変更	a3b2a46	beforeisland <beforeisland@gmail.com>	2020/10/14 23:59

gacha1987 commented on rt_delete.php:
```diff
@@ -0,0 +1,36 @@
+            $del = $db->prepare('DELETE FROM posts WHERE id=?');
+            $del->execute(array($_REQUEST['rt_id']));
+        } else {
+            echo 'error:ログイン者とリツイート投稿者と一致しません';
```
exit()をつけてください。

beforeisland replied:
修正完了。本件、処理の追記忘れです。

fix: いいね・リツイート消去プログラムへエラー処理追加	d000130	beforeisland <beforeisland@gmail.com>	2020/10/15 0:04

gacha1987 reviewed (commented):
コメントしました。
インデントがずれている箇所が散見されますのでもう一度見直してタブとスペースの統一をしてください。

beforeisland committed:
fix: インデント設定変更

PHPフォーマッター使用及びVS code設定変更による字下げの統一化。space4文字仕様。

beforeisland committed:
fix: いいね・リツイート消去プログラムへエラー処理追加

想定外のURLパラメータが入力された場合のエラー処理について、exit();による強制終了が入っていなかったため追加。

beforeisland committed:
add: いいね追加機能にいいね対象投稿の存在有無の検知機能を追加

バリデーション対策にて機能追加。SQLで投稿を検索し、issetで変数がセットされているか調べる。

beforeisland committed:
add: リツイート追加機能にリツイート対象投稿の存在有無の検知機能を追加

バリデーション対策にて機能追加。SQLで投稿を検索し、issetで変数がセットされているか調べる。

beforeisland committed:
update: バックアップしたデータベースのファイルを更新

バックアップファイルとして残してあるデータベースを更新していたのでコミット。

choimake commented on good_insert.php:
```diff
@@ -0,0 +1,39 @@
+//SQL:いいね対象の投稿が存在するか探し出す
+$postSearches = $db->prepare('SELECT * FROM posts WHERE id=?');
+$postSearches->execute(array($_REQUEST['post_id']));
+$postSearch = $postSearches->fetch();
```
ここも`post_id`のバリデーションが必要です。

beforeisland replied:
修正完了。post_idを実使用する前にバリデーションを行うよう処理順序を変更しました。

単にエラーを表示するためのバリデーションではなく、バリデーションにより安全と確認されたパラメータを使用してプログラムの処理を進めていくコーディングの必要性を認識しました。

fix: いいね・リツイート対象投稿を探し出すSQLへバリデーション処理追加	d8dddb7	beforeisland <beforeisland@gmail.com>	2020/10/15 23:05

choimake commented on rt_insert.php:
```diff
@@ -0,0 +1,40 @@
+////SQL:リツイート対象の投稿が存在するか探し出す
+$rtSearches = $db->prepare('SELECT * FROM posts WHERE id=?');
+$rtSearches->execute(array($_REQUEST['post_id']));
+$rtSearch = $rtSearches->fetch();
```
ここも`post_id`のバリデーションが必要です。

beforeisland replied:
修正完了。post_idを実使用する前にバリデーションを行うよう処理順序を変更しました。

fix: いいね・リツイート対象投稿を探し出すSQLへバリデーション処理追加	d8dddb7	beforeisland <beforeisland@gmail.com>	2020/10/15 23:05

choimake commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+images/
```
`member_pictures`以下の画像をignoreに設定してください。
（ユーザーによって追加される画像はバージョン管理不要です。）

逆に`images`は入れて良いです。
（アプリ本体のUIで使っている、かつサイズが大きくないため）

beforeisland replied:
修正完了。ユーザー所掌の部分はバージョン管理不要、アプリで使っているものはバージョン管理必要と認識しました(アイコンの素材を変えたい時などにアプリとしてバージョン管理の必要性がある)。

fix: gitignoreリスト更新	3d40fce	beforeisland <beforeisland@gmail.com>	2020/10/15 23:14

choimake replied:
> アプリで使っているものはバージョン管理必要

この辺はプロジェクトによるので、まぁ参考程度という感じで認識しておいてください。

beforeisland committed:
fix: いいね・リツイート対象投稿を探し出すSQLへバリデーション処理追加

URLパラメータの検査を行う前にSQLで使うデータとして使っていたリクエストを、バリデーション実施後に処理を行う仕様へ変更。

beforeisland committed:
fix: gitignoreリスト更新

imegesディレクトリ内の監視を行うために無視リストから外した。ユーザー管理のmember_pictureディレクトリ内の画像は無視リスト追加。

choimake commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+member_picture/
+
```
ここでは「member_pictureのディレクトリだけをコミットして欲しい」というお題のつもりでしたが、現在の設定ではそれができていません。

「gitは空のディレクトリをコミットできない」という性質がありまして、本来はこのままだとmember_pictureはリポジトリに追加されない状態になっています。
※今回は20201004084051izakaya.jpgなどがコミットされているため、member_pictureもコミットできています。

なので空のmember_pictureディレクトリのみをコミットするような設定に書き換えてください。

beforeisland replied:
.gitkeepファイルを作成し、ディレクトリは管理・中身の画像は無視するように変更しました。また、無視リストに追加しただけではコミットに残ることを認識しましたので、一旦画像ファイルは削除しました。

ディレクトリそのものはシステム側の構成物のため管理を行い、ユーザー所掌の画像ファイルは管理の必要性が薄いということですね。

fix: gitignore再構築	b17e614	beforeisland <beforeisland@gmail.com>	2020/10/17 11:26

choimake reviewed (commented):
コメントしましたので確認お願いします。

beforeisland committed:
add: member_pictureディレクトリをgit管理対象に追加

member_pictureディレクトリ内に.gitkeepファイルを作成し、ディレクトリのみ監視対象とした。

beforeisland committed:
fix: member_pictureディレクトリ内のファイル削除

コミット対象から外すため一旦削除

beforeisland committed:
fix: gitignore再構築

無視リストを再度構築しました

choimake commented on good_insert.php:
```diff
@@ -0,0 +1,39 @@
+require('dbconnect.php');
+
+//if:セッションに数値形式で値が保存されているか調査 
+if (is_numeric($_SESSION['id'])) {
+
+    //if:リクエストにて受け取ったURLパラメータが数値形式で存在しているか調査
+    if (is_numeric($_REQUEST['post_id'])) {
+
+        //SQL:いいね対象の投稿が存在するか探し出す
+        $postSearches = $db->prepare('SELECT * FROM posts WHERE id=?');
+        $postSearches->execute(array($_REQUEST['post_id']));
+        $postSearch = $postSearches->fetch();
+
+        //if:いいね対象の投稿が存在すればいいね追加
+        if (isset($postSearch['id'])) {
+            //いいねを追加する
+            $insert = $db->prepare('INSERT INTO goods SET member_id=?, post_id=?, created=NOW()');
+            $insert->execute(array(
+                $_SESSION['id'],
+                $_REQUEST['post_id']
+            ));
+        } else {
+            echo 'error:いいね対象の投稿が存在しません';
+            exit();
+        }
+    } else {
+        echo 'error:URLパラメータが不正です';
+        exit();
+    }
+} else {
+    echo 'error:セッションにidが保存されていません';
+    exit();
+}
```
(修正不要)
この辺りのコードが色々条件分岐が重なって、ネストが深い状態になっています。
3重程度でそんなに長い処理でもないのでまだ読み取れますが、どんどん処理が複雑になった際にこのままの書き方だとコードの可読性がどんどん下がっていきます。

なのでこういう場合はアーリーリターンとう手法を使って、なるべく深いネストを作らないようにする書き方がオススメです。

適応すると以下のような書き方になります。
```
if(!is_numeric($_SESSION['id'])){
    exit;
}

if(!is_numeric($_REQUEST['post_id'])){
    exit;
}

if(!isset($postSearch['id'])){
    exit;
}

insert->execute();
```

参考記事を載せとくので、暇な時に確認してみてください。
https://qiita.com/nskydiving/items/1076c411b002b0a3aec9

choimake commented on good_insert.php:
```diff
@@ -0,0 +1,39 @@
+if (is_numeric($_SESSION['id'])) {
+
+    //if:リクエストにて受け取ったURLパラメータが数値形式で存在しているか調査
+    if (is_numeric($_REQUEST['post_id'])) {
```
(修正不要）
is_numericの判定だけだと、マイナスの値、0などのような値も通ってしまうので、システム的にその値をSQLに入れたくないということであれば、その値も弾くようにした方が良いです。

choimake reviewed (approved):
LGTMです。
参考程度にコメントを書いてあるので、時間があるときに確認しておいてください。
お疲れ様でした。
